src/relay/pass/anf.cc

```diff
@@ -0,0 +1,143 @@
+             const Map<Expr, Var>& expr_to_var) :
+    var_to_expr_(var_to_expr), var_to_var_(var_to_var), expr_to_var_(expr_to_var) { }
+
+  /* while it seems like a good idea to override VisitExpr to push into letlist once for all,
+   * at quite a few case we dont want that (let, param, localvar, globalvar)
+   */
+
+  Expr Compound(const Expr& original, const Expr& e) {
+    Var v = expr_to_var_->count(original) != 0 ? expr_to_var_->at(original) : VarNode::make("x");
+    return ll.Push(v, e);
+  }
+
+  Expr VisitExpr_(const ConstantNode* op) final {
+    return Compound(GetRef<Constant>(op), GetRef<Constant>(op));
+  }
+
+  Expr VisitExpr_(const TupleNode* op) final {
+    std::vector<Expr> f;
+    for (const Expr& e : op->fields) {
+      f.push_back(VisitExpr(e));
+    }
+    return Compound(GetRef<Tuple>(op), TupleNode::make(f));
+  }
+
+  Expr VisitExpr_(const TupleGetItemNode* op) final {
+    return Compound(GetRef<TupleGetItem>(op),
+                    TupleGetItemNode::make(VisitExpr(op->tuple), op->index));
+  }
+
+  Expr VisitExpr_(const VarNode* op) final {
+    Var var = GetRef<Var>(op);
+    if (var_to_expr_->count(var) != 0) {
+      return var_to_expr_->at(var);
+    }
+    if (var_to_var_->count(var) != 0) {
+      return var_to_var_->at(var);
+    }
+    return var;
+  }
+
+  Expr VisitExpr_(const GlobalVarNode* op) final {
+    return GetRef<GlobalVar>(op);
+  }
+
+  Expr VisitExpr_(const ParamNode* op) final {
+    return GetRef<Param>(op);
+  }
+
+  Expr VisitExpr_(const FunctionNode* op) final {
+    return Compound(GetRef<Function>(op),
+                    FunctionNode::make(op->params,
+                                       op->ret_type,
+                                       ANF(var_to_expr_, var_to_var_, expr_to_var_, op->body),
+                                       op->type_params));
+  }
+
+  Expr VisitExpr_(const CallNode* op) final {
+    std::vector<Expr> args;
+    for (const Expr& e : op->args) {
+      args.push_back(VisitExpr(e));
+    }
+    return Compound(GetRef<Call>(op),
+                    CallNode::make(VisitExpr(op->op), args, op->attrs, op->type_args));
+  }
+
+  Expr VisitExpr_(const LetNode* op) final {
+    Var x = VarNode::make("x");
+    var_to_var_->insert(std::pair<Var, Var>(op->var, x));
+    expr_to_var_->insert(std::pair<Expr, Var>(op->value, x));
+    Expr val = VisitExpr(op->value);
+    var_to_expr_->insert(std::pair<Var, Expr>(op->var, val));
+    return VisitExpr(op->body);
+  }
+
+  Expr VisitExpr_(const IfNode* op) final {
+    return Compound(GetRef<If>(op),
+                    IfNode::make(VisitExpr(op->cond),
+                                 ANF(var_to_expr_, var_to_var_, expr_to_var_, op->true_branch),
+                                 ANF(var_to_expr_, var_to_var_, expr_to_var_, op->false_branch)));
+  }
+
+  Expr VisitExpr_(const OpNode* op) final {
+    return GetRef<Op>(op);
+  }
+
+ public:
+  static Expr ANF(const Map<Var, Expr>& var_to_expr,
```

Comment:
The additional level of indirection is not necessary, directly inline into caller.